src/main/java/net/minecraftforge/event/world/RaidEvent.java

```diff
@@ -0,0 +1,219 @@
+        }
+    }
+
+    /**
+     * RaidEvent.Load is triggered when
+     * {@link net.minecraft.world.entity.raid.Raids#load(ServerLevel, CompoundTag)}
+     * trying to load, users are able
+     * to retrieve custom raids here
+     */
+    public static class Load extends RaidEvent
+    {
+        private BiFunction<ServerLevel, CompoundTag, Raid> raidConstructor = (level, tag) -> null;
+
+        public Load()
+        {
+            super(null);
+        }
+
+        /**
+         * Sets a custom raid loader
+         * @param raidConstructor own raid constructor
+         */
+        public void setRaidConstructor(BiFunction<ServerLevel, CompoundTag, Raid> raidConstructor)
+        {
+            this.raidConstructor = raidConstructor;
+        }
+
+        public BiFunction<ServerLevel, CompoundTag, Raid> getRaidConstructor()
+        {
+            return raidConstructor;
+        }
+    }
+
+    /**
+     * RaidEvent.SpawnGroup is triggered when
+     * a raid is trying to spawn a new wave
+     *
+     * Users are allowed to change the spawn position,
+     * and force spawn a bonus group
+     *
+     * This event is {@link Cancelable}
+     * If cancelled, the group will be counted spawned
+     * like usual but the entities will not be added.
+     */
+    @Cancelable
+    public static class SpawnGroup extends RaidEvent
+    {
+        private BlockPos pos;
+        private boolean shouldSpawnBonusGroup;
+        public SpawnGroup(Raid raid, BlockPos pos, boolean shouldSpawnBonusGroup)
+        {
+            super(raid);
+            this.pos = pos;
+            this.shouldSpawnBonusGroup = shouldSpawnBonusGroup;
+        }
+
+        /**
+         * Gets the position raiders spawning on
+         * @return the spawn position
+         */
+        public BlockPos getPos()
+        {
+            return pos;
+        }
+
+        /**
+         * Sets the position of raiders going to spawn on
+         * @param pos the new spawn position
+         */
+        public void setPos(BlockPos pos)
+        {
+            this.pos = pos;
+        }
+
+        /**
+         * Should the raid spawn bonus groups
+         * Default {@link Raid#shouldSpawnBonusGroup()}
+         */
+        @SuppressWarnings("JavadocReference")
+        public boolean shouldSpawnBonusGroup()
+        {
+            return shouldSpawnBonusGroup;
+        }
+
+        /**
+         * Set whether the bonus group should spawn
+         */
+        public void setShouldSpawnBonusGroup(boolean shouldSpawnBonusGroup)
+        {
+            this.shouldSpawnBonusGroup = shouldSpawnBonusGroup;
+        }
+    }
+
+    /**
+     * RaidEvent.Stop is triggered when the raid
+     * attempts to end
+     *
+     * Users are allowed to prevent the raid
+     * for being cancelled
+     *
+     * This event is {@link Cancelable}
+     * If cancelled, the raid is not stopped
+     */
+    @Cancelable
+    public static class Stop extends RaidEvent
```

Comment:
A better name for this would be ``Ending`` as it would be more obvious what is happening and what canceling it causes.